radicle/src/cob/issue.rs

```diff
@@ -0,0 +1,766 @@
+    pub fn timestamp(&self) -> Timestamp {
+        self.timestamp
+    }
+}
+
+impl FromHistory for Issue {
+    fn type_name() -> &'static TypeName {
+        &TYPENAME
+    }
+
+    fn from_history(history: &History) -> Result<Self, Error> {
+        let doc = history.traverse(Automerge::new(), |mut doc, entry| {
+            match entry.contents() {
+                Contents::Automerge(bytes) => {
+                    match automerge::Change::from_bytes(bytes.clone()) {
+                        Ok(change) => {
+                            doc.apply_changes([change]).ok();
+                        }
+                        Err(_err) => {
+                            // Ignore
+                        }
+                    }
+                }
+            }
+            ControlFlow::Continue(doc)
+        });
+        let issue = Issue::try_from(doc)?;
+
+        Ok(issue)
+    }
+}
+
+impl TryFrom<&History> for Issue {
+    type Error = Error;
+
+    fn try_from(history: &History) -> Result<Self, Self::Error> {
+        Issue::from_history(history)
+    }
+}
+
+impl TryFrom<Automerge> for Issue {
+    type Error = DocumentError;
+
+    fn try_from(doc: Automerge) -> Result<Self, Self::Error> {
+        let doc = Document::new(&doc);
+        let obj_id = doc.get_id(automerge::ObjId::Root, "issue")?;
+        let title = doc.get(&obj_id, "title")?;
+        let comment_id = doc.get_id(&obj_id, "comment")?;
+        let author = doc.get(&obj_id, "author").map(Author::new)?;
+        let state = doc.get(&obj_id, "state")?;
+        let timestamp = doc.get(&obj_id, "timestamp")?;
+
+        let comment = shared::lookup::comment(doc, &comment_id)?;
+        let discussion: Discussion = doc.list(&obj_id, "discussion", shared::lookup::thread)?;
+        let labels: HashSet<Label> = doc.keys(&obj_id, "labels")?;
+
+        Ok(Self {
+            title,
+            state,
+            author,
+            comment,
+            discussion,
+            labels,
+            timestamp,
+        })
+    }
+}
+
+pub struct IssueStore<'a> {
+    store: Store<'a, Issue>,
+}
+
+impl<'a> IssueStore<'a> {
+    /// Create a new issue store.
+    pub fn new(store: Store<'a, Issue>) -> Self {
+        Self { store }
+    }
+
+    /// Get an issue by id.
+    pub fn get(&self, id: &ObjectId) -> Result<Option<Issue>, Error> {
+        self.store.get(id)
+    }
+
+    /// Create an issue.
+    pub fn create<G: Signer>(
+        &self,
+        title: &str,
+        description: &str,
+        labels: &[Label],
+        signer: &G,
+    ) -> Result<IssueId, Error> {
+        let author = self.store.author();
+        let timestamp = Timestamp::now();
+        let contents = events::create(&author, title, description, timestamp, labels)?;
+        let cob = self.store.create("Create issue", contents, signer)?;
+
+        Ok(*cob.id())
+    }
+
+    /// Remove an issue.
+    pub fn remove<G: Signer>(&self, _issue_id: &IssueId) -> Result<(), Error> {
+        todo!()
```

Comment:
Was this by mistake?

Reply:
Nah, I guess it just got copied over; it's a placeholder until we implement removing objects.